libevmasm/RuleList.h

```diff
@@ -697,6 +699,28 @@ std::vector<SimplificationRule<Pattern>> evmRuleList(
 			[]() -> Pattern { return Instruction::SELFBALANCE; }, false
 		});
 
+	// @returns i if _x == 2**i
+	auto binaryLogarithm = [](Word const& _x) -> std::optional<Word> {
+		for (size_t i = 0; i < Pattern::WordSize; i++)
+			if (_x == (Word(1) << i))
+				return Word(i);
+		return {};
+	};
+	if (_evmVersion.hasBitwiseShifting())
+	{
+		rules.emplace_back(
+			Builtins::EXP(2, X),
+			[=]() -> Pattern { return Builtins::SHL(X, 1); },
+			false
+		);
+		rules.emplace_back(
+			Builtins::EXP(A, X),
+			[=]() -> Pattern { return Builtins::SHL(Builtins::MUL(X, *binaryLogarithm(A.d())), 1); },
```

Comment:
Shouldn't this be

```
			[=]() -> Pattern { return Builtins::SHL(X, *binaryLogarithm(A.d()))); },
```

Reply:
If `MUL(X, *binaryLog(A.d()))` overflows to `u256` zero e.g., `2**256`, then `EXP(A, X)` will produce a different result than the simplified version which would left shift `1` by zero, producing `1` as the result.

The fuzzer has only managed to trigger this with large literal constants as exponents e.g., "0" like so:

```
mstore(0, exp(exp(2, 16), "0"))
```

Reply:
Ah, finally managed to come up with a string that when used as an exponent produces a different result post optimisation because the multiplication overflows in such a way that the left shift is `< 256`

Example:

```
{
  let e := "z\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01"
  mstore(0, exp(exp(2, 128), e))
}
```

Reply:
In summary, I feel one may need larger than 256 precision multiplication routine that accepts `X` and  `binaryLog(A)` as input and returns a pair `<bool, u256>` (overflowed, product) and if `overflowed` is true, then we either optimise the exponentiation to zero or does not apply the optimisation at all. If `overflowed` is false, then we apply the rule always.